gemini/sample-apps/llamaindex-rag/common/utils.py

```diff
@@ -0,0 +1,248 @@
+        if i > 0:  # Not the first node
+            previous_node = node_list[i - 1]
+            current_node.relationships[NodeRelationship.PREVIOUS] = RelatedNodeInfo(
+                node_id=previous_node.node_id
+            )
+
+        if i < len(node_list) - 1:  # Not the last node
+            next_node = node_list[i + 1]
+            current_node.relationships[NodeRelationship.NEXT] = RelatedNodeInfo(
+                node_id=next_node.node_id
+            )
+    return node_list
+
+
+def create_pdf_blob_list(bucket_name, prefix):
+    """
+    Create a list of Blob objects for processing.
+    """
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+    blobs = bucket.list_blobs(prefix=prefix)
+    logger.info(blobs)
+    return [
+        Blob(
+            path=f"gs://{bucket_name}/{blob.name}",
+            mimetype=blob.content_type or "application/pdf",
+        )
+        for blob in blobs
+        if blob.name.lower().endswith(".pdf")
+    ]
+
+
+def upload_directory_to_gcs(local_dir_path: str, bucket_name: str, prefix: str):
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+
+    for root, dirs, files in os.walk(local_dir_path):
+        for file in files:
+            local_file_path = os.path.join(root, file)
+            relative_path = os.path.relpath(local_file_path, local_dir_path)
+            gcs_blob_name = f"{prefix}/{relative_path}"
+
+            blob = bucket.blob(gcs_blob_name)
+            blob.upload_from_filename(local_file_path)
+            print(f"File {local_file_path} uploaded to {gcs_blob_name}")
+
+
+def ensure_directory_exists(directory):
+    """
+    Ensure that the given directory exists, creating it if necessary.
+    """
+    os.makedirs(directory, exist_ok=True)
+
+
+def clean_text(text):
+    """
+    Clean and preprocess the extracted text.
+    """
+    import re
+
+    # Remove extra whitespace
+    text = re.sub(r"\s+", " ", text).strip()
+
+    # Remove any non-printable characters
+    text = "".join(char for char in text if char.isprintable() or char.isspace())
+    print(f"Cleaned text length: {len(text)}")
+
+    return text
+
+
+def chunk_text(text, chunk_size=512, overlap=50):
+    """
+    Split the text into chunks of approximately equal size.
+    """
+    words = text.split()
+    chunks = []
+
+    for i in range(0, len(words), chunk_size - overlap):
+        chunk = " ".join(words[i : i + chunk_size])
+        chunks.append(chunk)
+
+    return chunks
+
+
+def create_text_nodes(chunks, metadata, chunk_size=512, overlap=50):
+    """
+    Create TextNode objects from text chunks.
+    """
+    nodes = []
+    for i, chunk in enumerate(chunks):
+        node = TextNode(
+            text=chunk,
+            metadata={
+                "chunk_id": i,
+                "chunk_size": chunk_size,
+                "chunk_overlap": overlap,
+                **metadata,
+            },
+        )
```

Comment:
Shouldn't there be built in methods in LlamaIndex for this?

Reply:
Fixed in [commit](https://github.com/GoogleCloudPlatform/generative-ai/commit/e0dd2113cfb85f68d8d9ddad31c2317921c52a85)